core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -761,30 +779,62 @@ public void validateBlocks(Function<Long, Boolean> validator, boolean repair)
-      try (JournalContext journalContext = createJournalContext()) {
-        // This must be flushed while holding the lock on mBlockContainerIdGenerator, in order to
-        // prevent subsequent calls to return ids that have not been journaled and flushed.
-        journalContext.append(getContainerIdJournalEntry());
+    long containerId = mBlockContainerIdGenerator.getNewContainerId();
+    if (containerId >= (mJournaledNextContainerId - mContainerIdReservationSize / 2)) {
+      if (containerId >= mJournaledNextContainerId) {
+        synchronized (mBlockContainerIdGenerator) {
+          // This container id is not safe with respect to the last journaled container id.
+          // Therefore, journal the new state of the container id. This implies that when a master
+          // crashes, the container ids within the reservation which have not been used yet will
+          // never be used. This is a tradeoff between fully utilizing the container id space, vs.
+          // improving master scalability.
+
+          // Set the next id to journal with a reservation of container ids, to avoid having to
+          // write to the journal for ids within the reservation.
+          long possibleMaxContainerId = mBlockContainerIdGenerator.getNextContainerId();
+          if (possibleMaxContainerId >= mJournaledNextContainerId) {
+            mJournaledNextContainerId = possibleMaxContainerId + mContainerIdReservationSize;
+            try (JournalContext journalContext = createJournalContext()) {
+              // This must be flushed while holding the lock on mBlockContainerIdGenerator, in
+              // order to prevent subsequent calls to return ids that have not been journaled
+              // and flushed.
+              journalContext.append(getContainerIdJournalEntry());
+            }
+          }
+        }
+      } else {
+        if (mContainerIdDetectorIsIdle) {
+          synchronized (mBlockContainerIdGenerator) {
+            if (mContainerIdDetectorIsIdle) {
+              mContainerIdDetectorIsIdle = false;
+              mContainerIdDetector.submit(() -> {
+                try {
+                  synchronized (mBlockContainerIdGenerator) {
+                    long possibleMaxContainerId = mBlockContainerIdGenerator.getNextContainerId();
+
+                    if (possibleMaxContainerId
+                        >= (mJournaledNextContainerId - mContainerIdReservationSize / 2)) {
+                      mJournaledNextContainerId = possibleMaxContainerId
+                          + mContainerIdReservationSize;
+                      try (JournalContext journalContext = createJournalContext()) {
+                        journalContext.append(getContainerIdJournalEntry());
+                      }
+                    }
+                  }
+                } catch (UnavailableException e) {
+                  LOG.error("Container Id Detector failed", e);
+                  throw new RuntimeException(e);
```

Comment:
I think it's better just to log the error and not rethrow it as a runtime exception, this will also happen when transitioning from primary master so we probably don't want to throw a runtime exception. Since it will still do the journal append call in the normal thread when it runs out of ids, it should be ok.

Reply:
Do you mean to remove rethrow in that method?